File: wagerproof-ios-native/WagerproofKit/Sources/WagerproofServices/RevenueCatService.swift
import AppTrackingTransparency
import Foundation
import RevenueCat
import WagerproofModels

/// `RevenueCatService` wraps the RevenueCat Swift SDK. It mirrors
/// `wagerproof-mobile/services/revenuecat.ts` byte-for-byte where APIs overlap
/// (configure / logIn / logOut / getCustomerInfo / getOfferings / restore).
///
/// SDK identity: `Purchases.shared` is the singleton entry point. The native
/// Swift SDK takes the same API keys (iOS = `appl_TFQYZRtHkCBrnaILkniTjsulyHK`)
/// and the same entitlement identifier (`"WagerProof Pro"`) as the RN bridge.
///
/// Initialization flow:
///   1. App launch → `bootstrap(userId: nil)` configures the SDK.
///   2. Auth completes → `logIn(userId:)` aliases the anonymous RC user to the
///      authenticated `user.id`.
///   3. Auth sign-out → `logOut()` resets to an anonymous RC user.
public final class RevenueCatService: @unchecked Sendable {
    public static let shared = RevenueCatService()

    public static let apiKey = "appl_TFQYZRtHkCBrnaILkniTjsulyHK"
    public static let entitlementIdentifier = "WagerProof Pro"

    // Mirrors RN `PAYWALL_PLACEMENTS` — these identifiers must match the
    // placements configured in the RevenueCat dashboard.
    public enum Placement {
        public static let onboarding = "onboarding"
        public static let genericFeature = "generic_feature"
        public static let agentFeature = "agent_feature"
        public static let tierUpgradePremium = "tier_upgrade_premium"
        public static let tierUpgradePro = "tier_upgrade_pro"
    }

    private var configured = false
    private var subscriberIdentity: SubscriberIdentity?

    private struct SubscriberIdentity {
        let email: String?
        let displayName: String?
        let phoneNumber: String?
        let authProvider: String?
        let username: String?
        let accountCreatedAt: Date?
    }

    private init() {}

    /// Configure the RevenueCat SDK with the iOS API key. Idempotent.
    /// Optional `userId` aliases the RC user immediately (matches RN's
    /// `initializeRevenueCat(userId)` shape).
    public func bootstrap(userId: String? = nil) {
        guard !configured else { return }
        // Lowercased for the same case-sensitivity reason as logIn below.
        let config = Configuration.Builder(withAPIKey: Self.apiKey)
            .with(appUserID: userId?.lowercased())
            .build()
        Purchases.configure(with: config)
        #if DEBUG
        Purchases.logLevel = .debug
        #endif
        configured = true
        applyTrackingConsent()
    }

    /// Push every trustworthy signed-in identity field available to RevenueCat.
    /// Standard fields are forwarded to supported attribution integrations
    /// (including Meta), while the WagerProof fields remain useful for RC
    /// customer inspection and targeting.
    public func setSubscriberIdentity(
        email: String?,
        displayName: String?,
        phoneNumber: String?,
        authProvider: String?,
        username: String?,
        accountCreatedAt: Date?
    ) {
        guard configured else { return }
        let identity = SubscriberIdentity(
            email: email,
            displayName: displayName,
            phoneNumber: phoneNumber,
            authProvider: authProvider,
            username: username,
            accountCreatedAt: accountCreatedAt
        )
        subscriberIdentity = identity
        applySubscriberIdentity(identity)
    }

    private func applySubscriberIdentity(_ identity: SubscriberIdentity) {
        guard ATTrackingManager.trackingAuthorizationStatus == .authorized else { return }
        Purchases.shared.attribution.setEmail(Self.nonEmpty(identity.email))
        Purchases.shared.attribution.setDisplayName(Self.nonEmpty(identity.displayName))
        Purchases.shared.attribution.setPhoneNumber(Self.nonEmpty(identity.phoneNumber))

        var attributes: [String: String] = [:]
        if let authProvider = Self.nonEmpty(identity.authProvider) {
            attributes["wagerproof_auth_provider"] = authProvider
        }
        if let username = Self.nonEmpty(identity.username) {
            attributes["wagerproof_username"] = username
        }
        if let accountCreatedAt = identity.accountCreatedAt {
            attributes["wagerproof_account_created_at"] = ISO8601DateFormatter().string(from: accountCreatedAt)
        }
        if !attributes.isEmpty {
            Purchases.shared.attribution.setAttributes(attributes)
        }

        // Re-apply the Meta install identifier after RevenueCat has attached the
        // authenticated App User ID. This makes the linkage robust even when
        // Supabase auth and RevenueCat logIn finish in different orders.
        if let fbAnonymousID = MetaAnalyticsService.shared.anonymousID(), !fbAnonymousID.isEmpty {
            Purchases.shared.attribution.setFBAnonymousID(fbAnonymousID)
        }
        Purchases.shared.attribution.collectDeviceIdentifiers()
    }

    /// Purchase and entitlement management work regardless of ATT. Advertising
    /// identifiers and matching attributes are attached only with authorization.
    private func applyTrackingConsent() {
        guard configured else { return }
        if ATTrackingManager.trackingAuthorizationStatus == .authorized {
            if let subscriberIdentity { applySubscriberIdentity(subscriberIdentity) }
            Purchases.shared.attribution.collectDeviceIdentifiers()
            if let anonymousID = MetaAnalyticsService.shared.anonymousID() {
                Purchases.shared.attribution.setFBAnonymousID(anonymousID)
            }
        } else {
            // Clear attributes saved by older builds or a previous authorization.
            Purchases.shared.attribution.setEmail(nil)
            Purchases.shared.attribution.setDisplayName(nil)
            Purchases.shared.attribution.setPhoneNumber(nil)
            Purchases.shared.attribution.setAttributes([
                "$fbAnonId": "", "$idfa": "", "$idfv": "", "$ip": "",
                "wagerproof_auth_provider": "", "wagerproof_username": "",
                "wagerproof_account_created_at": ""
            ])
        }
    }

    public func refreshAttributionAfterTrackingAuthorization(isAuthorized: Bool) async {
        guard configured else { return }
        // Read the OS again rather than trusting a caller's cached choice.
        applyTrackingConsent()
        do {
            _ = try await Purchases.shared.syncAttributesAndOfferingsIfNeeded()
        } catch {
            #if DEBUG
            print("[RevenueCat] Failed to sync attribution after ATT: \(error.localizedDescription)")
            #endif
        }
    }

    public var isConfigured: Bool { configured }

    private static func nonEmpty(_ value: String?) -> String? {
        guard let value = value?.trimmingCharacters(in: .whitespacesAndNewlines),
              !value.isEmpty else {
            return nil
        }
        return value
    }

    /// Identify a known user by their Supabase user id. Returns the resulting
    /// `CustomerInfo` and whether RC created a brand-new customer (matches
    /// RN's `{ customerInfo, created }` shape).
    @discardableResult
    public func logIn(userId: String) async throws -> (customerInfo: CustomerInfo, created: Bool) {
        // RC app-user-ids are CASE-SENSITIVE. Web/RN/Android and every backend
        // grant use the lowercase Supabase uuid, but Swift's UUID.uuidString is
        // UPPERCASE — logging in with it created a second, entitlement-less RC
        // customer for every iOS user (paywall + locked pages for paying users,
        // incident 2026-07). Normalize here so every caller inherits the fix.
        let result = try await Purchases.shared.logIn(userId.lowercased())
        // AuthStore and RevenueCatStore observe the same Supabase event
        // independently. Re-applying here makes the destination deterministic
        // whether profile loading or RevenueCat logIn completes first.
        applyTrackingConsent()
        return (result.customerInfo, result.created)
    }

    /// Reset to an anonymous RC user. Called when Supabase auth signs out.
    public func logOut() async {
        subscriberIdentity = nil
        do {
            _ = try await Purchases.shared.logOut()
        } catch {
            // RN swallows the error (`console.warn`) — match behavior so a
            // sign-out doesn't block UI.
        }
    }

    public func customerInfo() async throws -> CustomerInfo {
        try await Purchases.shared.customerInfo()
    }

    public func currentOffering() async throws -> Offering? {
        let offerings = try await Purchases.shared.offerings()
        return offerings.current
    }

    /// Sync pending attributes before targeting. The SDK owns default-placement
    /// fallback; nil is an intentional No Offering, never a reason to use current.
    public func offering(forPlacement placementIdentifier: String) async throws -> Offering? {
        guard configured else { throw PlacementError.notConfigured }
        let userID = Purchases.shared.appUserID
        guard let offerings = try await Purchases.shared.syncAttributesAndOfferingsIfNeeded() else {
            throw PlacementError.missingOfferings
        }
        guard userID == Purchases.shared.appUserID else { throw PlacementError.identityChanged }
        return offerings.currentOffering(forPlacement: placementIdentifier)
    }

    public enum PlacementError: LocalizedError {
        case notConfigured, identityChanged, missingOfferings
        public var errorDescription: String? {
            "Subscription options could not be loaded for this account. Please try again."
        }
    }

    public func restorePurchases() async throws -> CustomerInfo {
        try await Purchases.shared.restorePurchases()
    }

    public func syncPurchases() async throws -> CustomerInfo {
        try await Purchases.shared.syncPurchases()
    }

    /// Returns `true` when the customer info contains the WagerProof Pro
    /// entitlement in `.active`. Matches RN's
    /// `customerInfo.entitlements.active[ENTITLEMENT_IDENTIFIER] !== undefined`.
    public func hasProEntitlement(_ info: CustomerInfo) -> Bool {
        info.entitlements.active[Self.entitlementIdentifier] != nil
    }

    public func subscriptionTier(_ info: CustomerInfo) -> SubscriptionTier? {
        SubscriptionTier.resolve(activeEntitlementIDs: Set(info.entitlements.active.keys))
    }

    public struct ServerSubscriptionAccess: Decodable, Sendable {
        public let subscriptionTier: SubscriptionTier?
        public let isTieredCustomer: Bool?
    }

    /// The authenticated resolver checks historical uppercase/anonymous RC
    /// identities. A legacy Pro plan on another identity must beat a new lower
    /// tier on the current SDK identity.
    public func serverSubscriptionAccess() async throws -> ServerSubscriptionAccess {
        let client = await MainSupabase.shared.client
        return try await client.functions.invoke("resolve-my-entitlement")
    }

    public func isTieredCustomer(_ info: CustomerInfo) -> Bool {
        SubscriptionTier.isTieredCustomer(productIDs: info.allPurchasedProductIdentifiers,
                                         entitlementIDs: Set(info.entitlements.all.keys))
    }

    private func highestEntitlement(_ info: CustomerInfo) -> EntitlementInfo? {
        guard let tier = subscriptionTier(info) else { return nil }
        return info.entitlements.active[tier.entitlementID]
    }

    /// Map an active entitlement's productIdentifier to a coarse subscription
    /// type. Matches RN's `getActiveSubscriptionType(customerInfo)`.
    public func activeSubscriptionType(_ info: CustomerInfo) -> String? {
        guard let entitlement = highestEntitlement(info) else { return nil }
        let productId = entitlement.productIdentifier.lowercased()
        if productId.contains("lifetime") { return "lifetime" }
        if productId.contains("annual") || productId.contains("yearly") { return "yearly" }
        if productId.contains("monthly") { return "monthly" }
        return nil
    }

    /// Returns the active entitlement's productIdentifier (used by the
    /// secret-settings "Check Offerings" debug action).
    public func activeProductIdentifier(_ info: CustomerInfo) -> String? {
        highestEntitlement(info)?.productIdentifier
    }

    public func activeExpirationDate(_ info: CustomerInfo) -> Date? {
        highestEntitlement(info)?.expirationDate
    }
}

File: wagerproof-android-native/core/services/src/main/kotlin/com/wagerproof/core/services/RevenueCatService.kt
package com.wagerproof.core.services

import android.app.Activity
import android.content.Context
import android.content.Intent
import com.revenuecat.purchases.CustomerInfo
import com.revenuecat.purchases.LogLevel
import com.revenuecat.purchases.Offering
import com.revenuecat.purchases.Package
import com.revenuecat.purchases.PurchaseParams
import com.revenuecat.purchases.PurchaseResult
import com.revenuecat.purchases.Purchases
import com.revenuecat.purchases.PurchasesConfiguration
import com.revenuecat.purchases.WebPurchaseRedemption
import com.revenuecat.purchases.awaitCustomerInfo
import com.revenuecat.purchases.awaitLogIn
import com.revenuecat.purchases.awaitLogOut
import com.revenuecat.purchases.awaitOfferings
import com.revenuecat.purchases.awaitPurchase
import com.revenuecat.purchases.awaitRestore
import com.revenuecat.purchases.awaitSyncPurchases
import com.revenuecat.purchases.interfaces.RedeemWebPurchaseListener
import com.wagerproof.core.shared.AppGroup
import com.wagerproof.core.shared.AppGroupKey
import java.util.Locale
import kotlinx.coroutines.CancellationException
import java.util.Date
import kotlin.coroutines.resume
import kotlinx.coroutines.suspendCancellableCoroutine

/**
 * RevenueCat SDK wrapper — port of iOS `RevenueCatService.swift`, which itself
 * mirrors RN `services/revenuecat.ts` (configure / logIn / logOut /
 * getCustomerInfo / getOfferings / restore).
 *
 * Lifecycle: launch → [bootstrap] (null user) → auth completes → [logIn] with
 * the Supabase user id → sign-out → [logOut] back to anonymous.
 */
object RevenueCatService {

    // Public client SDK key for the existing WagerProof Android RevenueCat app.
    // This is intentionally the same value used by the shipping RN client.
    const val REVENUECAT_API_KEY = "goog_cilRlGISDEjNmpNebMglZPXnPLb"

    const val ENTITLEMENT_IDENTIFIER = "WagerProof Pro"

    // Mirrors RN `PAYWALL_PLACEMENTS` — must match the RevenueCat dashboard.
    object Placement {
        const val ONBOARDING = "onboarding"
        const val GENERIC_FEATURE = "generic_feature"
        const val AGENT_FEATURE = "agent_feature"
    }

    data class LoginResult(val customerInfo: CustomerInfo, val created: Boolean)

    /** Signed-in identity forwarded to RevenueCat's attribution integrations. */
    internal data class SubscriberIdentity(
        /** Normalized Supabase uid that owns every PII field below. */
        val userId: String,
        val email: String?,
        val displayName: String?,
        val phoneNumber: String?,
        val authProvider: String?,
        val username: String?,
        val accountCreatedAt: String?,
    )

    @Volatile
    private var configured = false

    @Volatile
    private var subscriberIdentity: SubscriberIdentity? = null

    val isConfigured: Boolean get() = configured

    /**
     * Configure the SDK. Idempotent. Optional [userId] aliases the RC user
     * immediately (matches RN's `initializeRevenueCat(userId)` shape).
     */
    @Synchronized
    fun bootstrap(context: Context, userId: String? = null) {
        if (configured) return
        if (BuildFlags.isDebugBuild) {
            Purchases.logLevel = LogLevel.DEBUG
        }
        Purchases.configure(
            PurchasesConfiguration.Builder(context, REVENUECAT_API_KEY)
                .appUserID(userId)
                .build(),
        )
        configured = true
        // Best-effort device identifier collection — matches RN's fire-and-forget call.
        Purchases.sharedInstance.collectDeviceIdentifiers()
        // `setSubscriberIdentity` is allowed to race bootstrap. The user-id
        // guard inside apply makes this safe for both anonymous and identified
        // configuration paths.
        subscriberIdentity?.let(::applySubscriberIdentity)

        // Hand Meta's anonymous install ID to RevenueCat as the `$fbAnonId`
        // subscriber attribute. RevenueCat's server-side Meta integration sends
        // purchase/trial conversions to the Conversions API, and WITHOUT this ID
        // those server events cannot be joined back to the install — the only
        // attribution path that survives a device with no usable advertising id.
        applyFacebookAnonymousId()
    }

    /**
     * Push every trustworthy signed-in identity field to RevenueCat. The standard
     * fields (email / display name / phone) are forwarded to RC's attribution
     * integrations — including its server-side Meta CAPI events, which have
     * nothing to join on without them. The `wagerproof_*` attributes are for RC
     * customer inspection and targeting.
     *
     * Cached so [logIn] can re-apply after RevenueCat attaches the authenticated
     * App User ID — AuthStore and RevenueCatStore observe the same Supabase event
     * independently, so the ordering isn't deterministic.
     */
    fun setSubscriberIdentity(
        userId: String,
        email: String?,
        displayName: String?,
        phoneNumber: String?,
        authProvider: String?,
        username: String?,
        accountCreatedAt: String?,
    ) {
        val identity = SubscriberIdentity(
            userId = normalizeUserId(userId),
            email = email,
            displayName = displayName,
            phoneNumber = phoneNumber,
            authProvider = authProvider,
            username = username,
            accountCreatedAt = accountCreatedAt,
        )
        subscriberIdentity = identity
        if (configured) applySubscriberIdentity(identity)
    }

    private fun applySubscriberIdentity(identity: SubscriberIdentity) {
        val purchases = Purchases.sharedInstance
        // AuthStore and RevenueCatStore observe the same Supabase event on
        // independent coroutines. Never apply a payload merely because it is
        // cached: before/after a direct A → B transition the SDK may still be
        // attached to the other account.
        if (normalizeUserId(purchases.appUserID) != identity.userId) return
        purchases.setEmail(nonEmpty(identity.email))
        purchases.setDisplayName(nonEmpty(identity.displayName))
        purchases.setPhoneNumber(nonEmpty(identity.phoneNumber))

        val attributes = buildMap<String, String> {
            nonEmpty(identity.authProvider)?.let { put("wagerproof_auth_provider", it) }
            nonEmpty(identity.username)?.let { put("wagerproof_username", it) }
            nonEmpty(identity.accountCreatedAt)?.let { put("wagerproof_account_created_at", it) }
        }
        if (attributes.isNotEmpty()) {
            purchases.setAttributes(attributes)
        }

        // Re-apply after the authenticated App User ID is attached so the linkage
        // holds whichever of Supabase auth / RevenueCat logIn finishes first.
        applyFacebookAnonymousId()
        purchases.collectDeviceIdentifiers()
    }

    private fun applyFacebookAnonymousId() {
        val anonymousId = MetaAnalyticsService.anonymousId()
        if (!anonymousId.isNullOrBlank()) {
            Purchases.sharedInstance.setFBAnonymousID(anonymousId)
        }
    }

    private fun nonEmpty(value: String?): String? = value?.trim()?.takeIf { it.isNotEmpty() }

    internal fun matchingSubscriberIdentity(
        identity: SubscriberIdentity?,
        userId: String,
    ): SubscriberIdentity? = identity?.takeIf { it.userId == normalizeUserId(userId) }

    internal fun normalizeUserId(userId: String): String = userId.trim().lowercase(Locale.ROOT)

    /** Synchronous auth-boundary teardown; safe before RevenueCat's network logout. */
    fun clearSubscriberIdentity() {
        subscriberIdentity = null
    }

    /** Identify a known user by Supabase user id → `(customerInfo, created)`. */
    suspend fun logIn(userId: String): LoginResult {
        val result = Purchases.sharedInstance.awaitLogIn(userId)
        matchingSubscriberIdentity(subscriberIdentity, userId)?.let(::applySubscriberIdentity)
        return LoginResult(result.customerInfo, result.created)
    }

    /** Reset to an anonymous RC user on Supabase sign-out. Errors swallowed so sign-out never blocks UI (matches RN). */
    suspend fun logOut() {
        clearSubscriberIdentity()
        try {
            Purchases.sharedInstance.awaitLogOut()
        } catch (cancellation: CancellationException) {
            throw cancellation
        } catch (_: Exception) {
        }
    }

    suspend fun customerInfo(): CustomerInfo =
        Purchases.sharedInstance.awaitCustomerInfo()

    /**
     * The RevenueCat customer this device is currently acting as. Needed by the
     * app-to-web checkout link — a web purchase that doesn't carry this id lands
     * on a DIFFERENT customer and its entitlement never reaches the device.
     */
    val appUserId: String
        get() = Purchases.sharedInstance.appUserID

    /**
     * Drop the cached `CustomerInfo` so the next fetch hits the server.
     *
     * Only for the app-to-web return trip: that purchase completes on
     * RevenueCat's servers with nothing pushing it to the device, and the SDK
     * caches customer info for ~5 minutes. Used sparingly — invalidating for any
     * other reason just burns a network round trip.
     */
    fun invalidateCustomerInfoCache() {
        Purchases.sharedInstance.invalidateCustomerInfoCache()
    }

    suspend fun currentOffering(): Offering? =
        Purchases.sharedInstance.awaitOfferings().current

    /** Placement offering with fallback to `offerings.current` — mirrors RN's `getCurrentOfferingForPlacement`. */
    suspend fun offering(forPlacement: String): Offering? {
        val offerings = Purchases.sharedInstance.awaitOfferings()
        return offerings.getCurrentOfferingForPlacement(forPlacement) ?: offerings.current
    }

    /**
     * Buy a package through Play Billing.
     *
     * Needed by the custom paywall, which drives checkout itself instead of
     * handing the offering to RevenueCatUI's template. Throws
     * `PurchasesTransactionException` — callers must branch on its
     * `userCancelled` flag before treating the failure as an error.
     *
     * [activity] must be the CURRENT resumed Activity: Play Billing launches its
     * sheet on it, and a stale reference silently no-ops.
     */
    suspend fun purchase(activity: Activity, pkg: Package): PurchaseResult =
        Purchases.sharedInstance.awaitPurchase(PurchaseParams.Builder(activity, pkg).build())

    suspend fun restorePurchases(): CustomerInfo =
        Purchases.sharedInstance.awaitRestore()

    suspend fun syncPurchases(): CustomerInfo =
        Purchases.sharedInstance.awaitSyncPurchases()

    /**
     * Parse a RevenueCat Web Billing deep link. A null result means the intent
     * is unrelated to RevenueCat and should continue through normal routing.
     */
    fun webPurchaseRedemption(intent: Intent): WebPurchaseRedemption? =
        Purchases.parseAsWebPurchaseRedemption(intent)

    /** Redeem a Web Billing purchase and bridge RevenueCat's callback to coroutines. */
    suspend fun redeemWebPurchase(
        redemption: WebPurchaseRedemption,
    ): RedeemWebPurchaseListener.Result = suspendCancellableCoroutine { continuation ->
        Purchases.sharedInstance.redeemWebPurchase(redemption) { result ->
            if (continuation.isActive) continuation.resume(result)
        }
    }

    // Helpers — mirror RN's entitlement checks so gating stays 1:1 across platforms.

    fun hasProEntitlement(info: CustomerInfo): Boolean =
        info.entitlements.active.containsKey(ENTITLEMENT_IDENTIFIER)

    /** Coarse subscription type from the active entitlement's productIdentifier. */
    fun activeSubscriptionType(info: CustomerInfo): String? {
        val productId = info.entitlements.active[ENTITLEMENT_IDENTIFIER]
            ?.productIdentifier?.lowercase() ?: return null
        return when {
            productId.contains("lifetime") -> "lifetime"
            productId.contains("annual") || productId.contains("yearly") -> "yearly"
            productId.contains("monthly") -> "monthly"
            else -> null
        }
    }

    fun activeProductIdentifier(info: CustomerInfo): String? =
        info.entitlements.active[ENTITLEMENT_IDENTIFIER]?.productIdentifier

    fun activeExpirationDate(info: CustomerInfo): Date? =
        info.entitlements.active[ENTITLEMENT_IDENTIFIER]?.expirationDate

    /**
     * Persist the coarse entitlement snapshot so widgets/cold-launch UI don't
     * flash "free" while RC reconciles. On iOS this write lives in
     * `RevenueCatStore.apply()` after every accepted CustomerInfo — the future
     * :core:stores RevenueCatStore must call this at that same point.
     */
    fun persistEntitlementSnapshot(info: CustomerInfo) {
        val granted = hasProEntitlement(info)
        val type = activeSubscriptionType(info)
        val editor = AppGroup.prefs.edit()
        editor.putBoolean(AppGroupKey.PRO_ENTITLEMENT_GRANTED, granted)
        if (type != null) {
            editor.putString(AppGroupKey.PRO_SUBSCRIPTION_TYPE, type)
        } else {
            editor.remove(AppGroupKey.PRO_SUBSCRIPTION_TYPE)
        }
        editor.apply()
    }
}
